packages/dd-trace/src/appsec/iast/analyzers/header-injection-analyzer.js

```diff
@@ -0,0 +1,96 @@
+'use strict'
+
+const InjectionAnalyzer = require('./injection-analyzer')
+const { HEADER_INJECTION } = require('../vulnerabilities')
+const { getNodeModulesPaths } = require('../path-line')
+const { HEADER_NAME_VALUE_SEPARATOR } = require('../vulnerabilities-formatter/constants')
+const { getRanges } = require('../taint-tracking/operations')
+const {
+  HTTP_REQUEST_COOKIE_NAME,
+  HTTP_REQUEST_COOKIE_VALUE,
+  HTTP_REQUEST_HEADER_VALUE
+} = require('../taint-tracking/source-types')
+
+const EXCLUDED_PATHS = getNodeModulesPaths('express')
+const EXCLUDED_HEADER_NAMES = [
+  'location',
+  'sec-websocket-location',
+  'sec-websocket-accept',
+  'upgrade',
+  'connection'
+]
+
+class HeaderInjectionAnalyzer extends InjectionAnalyzer {
+  constructor () {
+    super(HEADER_INJECTION)
+  }
+
+  onConfigure () {
+    this.addSub('datadog:http:server:response:set-header:finish', ({ name, value }) => {
+      if (Array.isArray(value)) {
+        for (let i = 0; i < value.length; i++) {
+          const headerValue = value[i]
+
+          this.analyze({ name, value: headerValue })
+        }
+      } else {
+        this.analyze({ name, value })
+      }
+    })
+  }
+
+  _isVulnerable ({ name, value }, iastContext) {
+    const lowerCasedHeaderName = name?.trim().toLowerCase()
+
+    if (this.isExcludedHeaderName(lowerCasedHeaderName) || typeof value !== 'string') return
+
+    return super._isVulnerable(value, iastContext) &&
+      !(this.isCookieExclusion(lowerCasedHeaderName, value, iastContext) ||
```

Comment:
This conditional could potentially call `getRanges` twice (one for `isCookieExclusion` and one for `isAccessControlAllowOriginExclusion`). Could we try to get it once to avoid calling it twice?

Reply:
Not, it can't.
`isCookieExecution` calls to `getRanges` only if the header name is `set-cookie`, and `isAccessControlAllowOriginEclusion` calls to it only when the header name is `access-control-allow-origin`. But if the header name is any other value, they don't call to `getRanges`.